node/src/components/storage.rs

```diff
@@ -1074,56 +1083,67 @@ impl Storage {
     fn get_single_block_body_v2<Tx: Transaction>(
         &self,
         tx: &mut Tx,
         block_body_hash: &Digest,
     ) -> Result<Option<BlockBody>, LmdbExtError> {
-        let [hashed_deploy_hashes, hashed_transfer_hashes, hashed_proposer]: [Digest;
-            BlockBodyHashedParts::NUMBER_OF_HASHES] =
-            match tx.get_value(self.block_body_v2_db, block_body_hash)? {
-                Some(parts_hashes) => parts_hashes,
+        let deploy_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+            match self.get_merkle_linked_list_node(tx, self.deploy_hashes_db, block_body_hash)? {
+                Some(deploy_hashes_with_proof) => deploy_hashes_with_proof,
                 None => return Ok(None),
             };
-
-        let deploy_hashes: Vec<DeployHash> =
-            match tx.get_value(self.deploy_hashes_db, &hashed_deploy_hashes)? {
-                Some(deploy_hashes) => deploy_hashes,
-                None => {
-                    return Err(LmdbExtError::BlockBodyDidNotHaveDeployHashes {
-                        block_body_hash: *block_body_hash,
-                        hashed_deploy_hashes,
-                    })
-                }
-            };
-
-        let transfer_hashes: Vec<DeployHash> =
-            match tx.get_value(self.transfer_hashes_db, &hashed_transfer_hashes)? {
-                Some(transfer_hashes) => transfer_hashes,
-                None => {
-                    return Err(LmdbExtError::BlockBodyDidNotHaveTransferHashes {
-                        block_body_hash: *block_body_hash,
-                        hashed_transfer_hashes,
-                    })
-                }
-            };
-
-        let proposer: PublicKey = match tx.get_value(self.proposer_db, &hashed_proposer)? {
-            Some(proposer) => proposer,
-            None => {
-                return Err(LmdbExtError::BlockBodyDidNotHaveProposerHash {
-                    block_body_hash: *block_body_hash,
-                    hashed_proposer,
-                })
-            }
+        let transfer_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> = match self
+            .get_merkle_linked_list_node(
+                tx,
+                self.transfer_hashes_db,
+                deploy_hashes_with_proof.merkle_proof_of_rest(),
+            )? {
+            Some(transfer_hashes_with_proof) => transfer_hashes_with_proof,
+            None => return Ok(None),
         };
-
-        let block_body = BlockBody::new(proposer, deploy_hashes, transfer_hashes);
+        let proposer_with_proof: MerkleLinkedListNode<PublicKey> = match self
+            .get_merkle_linked_list_node(
+                tx,
+                self.proposer_db,
+                transfer_hashes_with_proof.merkle_proof_of_rest(),
+            )? {
+            Some(proposer_with_proof) => proposer_with_proof,
+            None => return Ok(None),
+        };
```

Comment:
Should we check that `proposer_with_proof.merkle_proof_of_rest() == SENTINEL` here? That might change in the future, of course, but it's expected for now, so maybe we should be checking it.

Reply:
Fixed in https://github.com/casper-network/casper-node/pull/1563/commits/45aaf924c790a5f620934cc7f00fd24a24601d02